resolver/resolver.go

```diff
@@ -67,77 +76,72 @@ func (r *Resolver) ResolveToLastNode(ctx context.Context, fpath path.Path) (cid.
 		return c, nil, nil
 	}
 
-	nd, err := r.DAG.Get(ctx, c)
+	// create a selector to traverse all path segments but only match the last
+	pathSelector, err := pathLeafSelector(p[:len(p)-1])
 	if err != nil {
 		return cid.Cid{}, nil, err
 	}
 
-	for len(p) > 0 {
-		lnk, rest, err := r.ResolveOnce(ctx, r.DAG, nd, p)
-
-		// Note: have to drop the error here as `ResolveOnce` doesn't handle 'leaf'
-		// paths (so e.g. for `echo '{"foo":123}' | ipfs dag put` we wouldn't be
-		// able to resolve `zdpu[...]/foo`)
-		if lnk == nil {
-			break
-		}
-
-		if err != nil {
-			if err == dag.ErrLinkNotFound {
-				err = ErrNoLink{Name: p[0], Node: nd.Cid()}
-			}
-			return cid.Cid{}, nil, err
-		}
-
-		if len(rest) == 0 {
-			return lnk.Cid, nil, nil
-		}
+	// resolve node before last path segment
+	nodes, err := r.resolveNodes(ctx, c, pathSelector)
+	if err != nil {
+		return cid.Cid{}, nil, err
+	}
+	if len(nodes) < 1 {
+		return cid.Cid{}, nil, fmt.Errorf("path %v did not resolve to a node", fpath)
+	}
+	parent := nodes[len(nodes)-1]
 
-		next, err := lnk.GetNode(ctx, r.DAG)
-		if err != nil {
-			return cid.Cid{}, nil, err
-		}
-		nd = next
-		p = rest
+	// find final path segment within node
+	nd, err := parent.LookupByString(p[len(p)-1])
+	if err != nil {
+		return cid.Cid{}, nil, err
 	}
 
-	if len(p) == 0 {
-		return nd.Cid(), nil, nil
+	if nd.Kind() != ipldp.Kind_Link {
+		return cid.Cid{}, nil, fmt.Errorf("path %v resolves to something other than a cid link: %v", fpath, nd)
```

Comment:
This error message doesn't really make sense to me, I'd propose something more like

```suggestion
		return cid.Cid{}, nil, fmt.Errorf("ResolveToLastNode expected path %v to resolve to a link kind, but got %v", fpath, nd.Kind())
```